crates/usvg/src/text/layout.rs

```diff
@@ -2,364 +2,465 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-use std::collections::HashMap;
-use std::num::NonZeroU16;
-use std::sync::Arc;
-
+use crate::tree::{BBox, IsValidLength};
+use crate::{
+    AlignmentBaseline, ApproxZeroUlps, BaselineShift, DominantBaseline, Fill, FillRule, Font,
+    FontStretch, FontStyle, LengthAdjust, PaintOrder, Path, ShapeRendering, Stroke, Text,
+    TextAnchor, TextChunk, TextDecorationStyle, TextFlow, TextPath, TextSpan, Visibility,
+    WritingMode,
+};
 use fontdb::{Database, ID};
 use kurbo::{ParamCurve, ParamCurveArclen, ParamCurveDeriv};
 use rustybuzz::ttf_parser;
+use rustybuzz::ttf_parser::GlyphId;
+use std::collections::HashMap;
+use std::num::NonZeroU16;
+use std::sync::Arc;
+use strict_num::NonZeroPositiveF32;
 use svgtypes::FontFamily;
-use ttf_parser::GlyphId;
+use tiny_skia_path::{NonZeroRect, Transform};
 use unicode_script::UnicodeScript;
 
-use crate::*;
+/// A glyph that has already been positioned correctly.
+///
+/// Note that the transform already takes the font size into consideration, so applying the
+/// transform to the outline of the glyphs is all that is necessary to display it correctly.
+#[derive(Clone, Copy, Debug)]
+pub struct PositionedGlyph {
+    /// The transform of the glyph. This transform should be applied to the _glyph outlines_, meaning
+    /// that paint servers referenced by the glyph's span should not be affected by it.
+    pub transform: Transform,
+    /// The ID of the glyph.
+    pub glyph_id: GlyphId,
+    /// The ID of the font the glyph should be taken from.
+    pub font: ID,
+}
+
+/// A span contains a number of layouted glyphs that share the same fill, stroke, paint order and
+/// visibility.
+#[derive(Clone, Debug)]
+pub struct Span {
+    /// The fill of the span.
+    pub fill: Option<Fill>,
+    /// The stroke of the span.
+    pub stroke: Option<Stroke>,
+    /// The paint order of the span.
+    pub paint_order: PaintOrder,
+    /// The font size of the span.
+    pub font_size: NonZeroPositiveF32,
+    /// The visibility of the span.
+    pub visibility: Visibility,
+    /// The glyphs that make up the span.
+    pub positioned_glyphs: Vec<PositionedGlyph>,
+}
 
-pub(crate) fn convert(text: &mut Text, fontdb: &fontdb::Database) -> Option<()> {
-    let (new_paths, bbox, stroke_bbox) = text_to_paths(text, fontdb)?;
+/// A text fragment is a component of a text element.
+///
+/// It can either be a span (which contains a number of layouted glyphs that share
+/// some properties) or a path (which is a layouted decoration span).
+#[derive(Clone, Debug)]
+pub enum TextFragment {
+    /// A text span.
+    Span(Span),
+    /// A path.
+    Path(Path),
```

Comment:
What is `Path` here? The `Path` node? Why?

Reply:
Yes, it's a usvg path. Right now it's basically just a decoration span. I guess I should rename the enum variant.

Reply:
Oh, so those are for underlines and stuff? Hm... good question about we suppose to store them.
I guess we need just a list of rectangles. Maybe something like:

```rust
struct Layout {
    fragments: Vec<TextFragment>,
    underlines: Vec<NonZeroRect>,
    overlines: Vec<TextFragment>,
    strikethrougs: Vec<TextFragment>,
}
```

Reply:
I'm not sure I understand, are you suggesting to store text fragments that have different decoration spans in different vectors? But that would mess up the order, no? What problem do you see with the current approach? By storing decorations as paths and text spans in the same vector, we can simply iterate over them in the order they are defined in the vector and render them separately so we don't need to worry about that.

Reply:
And it's also nice if I as a consumer of the API don't need to worry about resolving the text decorations myself but can simply render the resulting paths.

Reply:
Actually... What I could do is that we store the paths of underlines, line-throughs and overlines in the span itself. And then we basically only have a list of spans. This might be a bit cleaner. Do you want me to change this?

Reply:
My problem is that text decoration should not be part of `TextFragment`. Because it's kinda isn't.
If we have a underline over the whole text - then we have a single line/path. Why should it be part of `TextFragment `? It doesn't make any sense to me.

In case of text-on-path decorations _can_ be rotated, meaning just a Rect wouldn't be enough. So Path is fine for now.

And no, no one would be resolving text decoration manually, mainly because it would be impossible.

Reply:
As mentioned below, I've removed `TextFragment` and store text decorations with the span now, I hope that's better!